src/client/visualizations/line-chart/utils/calculate-x-scale.ts

```diff
@@ -0,0 +1,99 @@
+import { Timekeeper } from "../../../../common/models/timekeeper/timekeeper";
+import { union } from "../../../../common/utils/plywood/range";
+import { ContinuousScale } from "./scale";
+import { getContinuousDimension, getContinuousSplit } from "./splits";
+
+function getScaleX(kind: ContinuousDimensionKind, { start, end }: PlywoodRange, stageWidth: number): ContinuousScale {
+  const range = [0, stageWidth];
+  switch (kind) {
+    case "number": {
+      const domain = [start, end] as [number, number];
+      return d3.scale.linear().domain(domain).range(range);
+    }
+    case "time": {
+      const domain = [start, end] as [Date, Date];
+      return d3.time.scale().domain(domain).range(range);
+    }
+  }
+}
+
+function ensureMaxTime(axisRange: PlywoodRange, maxTime: Date, continuousSplit: Split, timezone: Timezone) {
+  // Special treatment for realtime data, i.e. time data where the maxTime is within Duration of the filter end
+  const continuousBucket = continuousSplit.bucket;
+  if (maxTime && continuousBucket instanceof Duration) {
+    const axisRangeEnd = axisRange.end as Date;
+    const axisRangeEndFloor = continuousBucket.floor(axisRangeEnd, timezone);
+    const axisRangeEndCeil = continuousBucket.shift(axisRangeEndFloor, timezone);
+    if (maxTime && axisRangeEndFloor < maxTime && maxTime < axisRangeEndCeil) {
+      return Range.fromJS({ start: axisRange.start, end: axisRangeEndCeil });
+    }
+  }
+  return axisRange;
+}
+
+function getFilterRange(essence: Essence, continuousSplit: Split, timekeeper: Timekeeper): PlywoodRange {
+  const maxTime = essence.dataCube.getMaxTime(timekeeper);
+  const continuousDimension = essence.dataCube.getDimension(continuousSplit.reference);
+  const effectiveFilter = essence
+    .getEffectiveFilter(timekeeper);
+  const continuousFilter = effectiveFilter.getClauseForDimension(continuousDimension);
+
+  let range = null;
+  if (continuousFilter instanceof NumberFilterClause) {
+    range = NumberRange.fromJS(continuousFilter.values.first());
+  }
+  if (continuousFilter instanceof FixedTimeFilterClause) {
+    range = TimeRange.fromJS(continuousFilter.values.first());
+  }
+  return ensureMaxTime(range, maxTime, continuousSplit, essence.timezone);
+}
+
+function getDatasetXRange(dataset: Dataset, continuousDimension: Dimension): PlywoodRange {
+  if (!dataset || dataset.count() === 0) return null;
+  const key = continuousDimension.name;
+
+  const firstDatum = dataset.data[0];
+  let ranges: PlywoodRange[];
+  if (firstDatum["SPLIT"]) {
+    ranges = dataset.data.map(d => getDatasetXRange(d["SPLIT"] as Dataset, continuousDimension));
+  } else {
+    ranges = dataset.data.map(d => (d as any)[key] as PlywoodRange);
+  }
+
+  return ranges.reduce((a: PlywoodRange, b: PlywoodRange) => (a && b) ? a.extend(b) : (a || b));
+}
+
+export default function calculateXScale(essence: Essence, timekeeper: Timekeeper, dataset: Dataset, width: number): ContinuousScale {
```

Comment:
Here be dragons. I just moved few methods from old code. Refactor in progress ... 🤖